util/linuxfw/iptables_runner.go

```diff
@@ -0,0 +1,485 @@
+	}
+	if i.v6NATAvailable {
+		if err := i.ipt6.Append("nat", "ts-postrouting", args...); err != nil {
+			return fmt.Errorf("adding %v in v6/nat/ts-postrouting: %w", args, err)
+		}
+	}
+	return nil
+}
+
+// DelSNATRule removes the netfilter rule to SNAT traffic destined for
+// local subnets. An error is returned if the rule does not exist.
+func (i *iptablesRunner) DelSNATRule() error {
+	args := []string{"-m", "mark", "--mark", TailscaleSubnetRouteMark + "/" + TailscaleFwmarkMask, "-j", "MASQUERADE"}
+	if err := i.ipt4.Delete("nat", "ts-postrouting", args...); err != nil {
+		return fmt.Errorf("deleting %v in v4/nat/ts-postrouting: %w", args, err)
+	}
+	if i.v6NATAvailable {
+		if err := i.ipt6.Delete("nat", "ts-postrouting", args...); err != nil {
+			return fmt.Errorf("deleting %v in v6/nat/ts-postrouting: %w", args, err)
+		}
+	}
+	return nil
+}
+
+// IptablesCleanUp clears the netfilter rules for IPv4 and IPv6.
+// If error occurs, it's logged but not returned.
+func IptablesCleanUp(logf logger.Logf) {
+	err := clearRules(iptables.ProtocolIPv4, logf)
+	if err != nil {
+		logf("ERROR: clear iptables: %w", err)
+	}
+
+	err = clearRules(iptables.ProtocolIPv6, logf)
+	if err != nil {
+		logf("ERROR: clear ip6tables: %w", err)
+	}
+}
+
+// delTSHook deletes hook in a chain that jumps to a ts-chain. If the hook does not
+// exist, it's a no-op since the desired state is already achieved but we log the
+// error because error code from the iptables module resists unwrapping.
+func delTSHook(ipt iptablesInterface, table, chain string, logf logger.Logf) error {
+	tsChain := tsChain(chain)
+	args := []string{"-j", tsChain}
+	if err := ipt.Delete(table, chain, args...); err != nil {
+		// TODO(apenwarr): check for errCode(1) here.
+		// Unfortunately the error code from the iptables
+		// module resists unwrapping, unlike with other
+		// calls. So we have to assume if Delete fails,
+		// it's because there is no such rule.
+		logf("deleting %v in %s/%s: %w", args, table, chain, err)
+		return nil
+	}
+	return nil
+}
+
+// delChain flushs and deletes a chain. If the chain does not exist, it's a no-op
+// since the desired state is already achieved. otherwise, it returns an error.
+func delChain(ipt iptablesInterface, table, chain string) error {
+	if err := ipt.ClearChain(table, chain); err != nil {
+		if isErrChainNotExist(err) {
+			// nonexistent chain. nothing to do.
+			return nil
+		}
+		return fmt.Errorf("flushing %s/%s: %w", table, chain, err)
+	}
+	if err := ipt.DeleteChain(table, chain); err != nil {
+		return fmt.Errorf("deleting %s/%s: %v", table, chain, err)
+	}
+	return nil
+}
+
+// clearRules clears all the iptables rules created by Tailscale
+// for the given protocol. If error occurs, it's logged but not returned.
+func clearRules(proto iptables.Protocol, logf logger.Logf) error {
+	ipt, err := iptables.NewWithProtocol(proto)
+	if err != nil {
+		return err
+	}
+
```

Comment:
consider using https://pkg.go.dev/tailscale.com/util/multierr here, and letting the parent log once, rather than producing many individual logs.